consensus/src/context/task.rs

```diff
@@ -0,0 +1,232 @@
+//! Context Task
+//!
+//! This module contains the async task that handles keeping track of blockchain context.
+//! It holds all the context caches and handles [`tower::Service`] requests.
+//!
+use futures::channel::oneshot;
+use tokio::sync::mpsc;
+use tower::ServiceExt;
+use tracing::Instrument;
+
+use cuprate_consensus_rules::blocks::ContextToVerifyBlock;
+
+use super::{
+    difficulty, hardforks, rx_vms, weight, BlockChainContext, BlockChainContextRequest,
+    BlockChainContextResponse, ContextConfig, RawBlockChainContext, ValidityToken,
+    BLOCKCHAIN_TIMESTAMP_CHECK_WINDOW,
+};
+use crate::{Database, DatabaseRequest, DatabaseResponse, ExtendedConsensusError};
+
+/// A request from the context service to the context task.
+pub(super) struct ContextTaskRequest {
+    /// The request.
+    pub req: BlockChainContextRequest,
+    /// The response channel.
+    pub tx: oneshot::Sender<Result<BlockChainContextResponse, tower::BoxError>>,
+    /// The tracing span of the requester.
+    pub span: tracing::Span,
+}
+
+/// The Context task that keeps the blockchain context and handles requests.
+pub struct ContextTask {
+    /// A token used to invalidate previous contexts when a new
+    /// block is added to the chain.
+    current_validity_token: ValidityToken,
+
+    /// The difficulty cache.
+    difficulty_cache: difficulty::DifficultyCache,
+    /// The weight cache.
+    weight_cache: weight::BlockWeightsCache,
+    /// The RX VM cache.
+    rx_vm_cache: rx_vms::RandomXVMCache,
+    /// The hard-fork state cache.
+    hardfork_state: hardforks::HardForkState,
+
+    /// The current chain height.
+    chain_height: u64,
+    /// The top block hash.
+    top_block_hash: [u8; 32],
+    /// The total amount of coins generated.
+    already_generated_coins: u64,
+}
+
+impl ContextTask {
+    /// Initialize the [`ContextTask`], this will need to pull a lot of data from the database so may take a
+    /// while to complete.
+    pub async fn init_context<D>(
+        cfg: ContextConfig,
+        mut database: D,
+    ) -> Result<ContextTask, ExtendedConsensusError>
+    where
+        D: Database + Clone + Send + Sync + 'static,
+        D::Future: Send + 'static,
+    {
+        let ContextConfig {
+            difficulty_cfg,
+            weights_config,
+            hard_fork_cfg,
+        } = cfg;
```

Comment:
Probably don't do this but this code made me wonder if this was possible, and it is:
```rust
    pub async fn init_context<D>(
        ContextConfig {
            difficulty_cfg,
            weights_config,
            hard_fork_cfg,
        }: ContextConfig,
        mut database: D,
    ) -> Result<ContextTask, ExtendedConsensusError>
    where
        D: Database + Clone + Send + Sync + 'static,
        D::Future: Send + 'static,
    {
        // already destructured
```

Reply:
ah interesting!